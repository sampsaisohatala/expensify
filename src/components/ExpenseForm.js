import React, { useState } from 'react';

const ExpenseForm = () => {
   const [description, setDescription] = useState('');
   const [note, setNote] = useState('');
   const [amount, setAmount] = useState('');

   const onDescriptionChange = (e) => {
      setDescription(e.target.value);
   };
   const onNoteChange = (e) => {
      setNote(e.target.value);
   };
   const onAmountChange = (e) => {
      console.log('onAmountChange');
      const amount = e.target.value;
      if (amount.match(/^\d*(\.\d{0,2})?$/)) {
         // regular expression -> https://regex101.com/
         setAmount(amount);
      }
   };

   return (
      <div>
         <h3>Expense form</h3>
         <form>
            <input type="text" placeholder="Description" autoFocus value={description} onChange={onDescriptionChange} />
            <input type="number" placeholder="Amount" value={amount} onChange={onAmountChange} />
            <textarea placeholder="Add a note to your expense (optional)" value={note} onChange={onNoteChange}></textarea>
            <button>Add expense</button>
         </form>
      </div>
   );
};

export default ExpenseForm;
