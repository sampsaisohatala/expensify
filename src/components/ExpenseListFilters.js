import React, { useState } from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

const ExpenseListFilters = (props) => {
   const [calenderFocused, setCalenderFocused] = useState(null);

   const onDatesChange = ({ startDate, endDate }) => {
      props.dispatch(setStartDate(startDate));
      props.dispatch(setEndDate(endDate));
   };

   const onFocusChange = (calenderFocused) => {
      setCalenderFocused(calenderFocused);
   };

   return (
      <div>
         {/* Text filter */}
         <input
            type="text"
            value={props.filters.text}
            onChange={(e) => {
               console.log(e.target.value);
               props.dispatch(setTextFilter(e.target.value));
            }}
         />
         {/* Sort by */}
         <select
            value={props.filters.sortBy}
            onChange={(e) => {
               if (e.target.value === 'date') {
                  props.dispatch(sortByDate());
               } else if (e.target.value === 'amount') {
                  props.dispatch(sortByAmount());
               }
            }}
         >
            <option value="date">Date</option>
            <option value="amount">Amount</option>
         </select>
         {/* Date filter */}
         <DateRangePicker
            startDate={props.filters.startDate}
            startDateId="your_unique_start_date_id"
            endDate={props.filters.endDate}
            endDateId="your_unique_end_date_id"
            onDatesChange={({ startDate, endDate }) => onDatesChange({ startDate, endDate })}
            focusedInput={calenderFocused}
            onFocusChange={(focusedInput) => onFocusChange(focusedInput)}
            showClearDates={true}
            numberOfMonths={1}
            firstDayOfWeek={1}
            isOutsideRange={() => false}
         />
      </div>
   );
};

const mapStateToProps = (state) => {
   return {
      filters: state.filters,
   };
};

export default connect(mapStateToProps)(ExpenseListFilters);
