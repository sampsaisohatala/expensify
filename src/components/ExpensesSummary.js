import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expensesCount, expensesTotalAmount }) => {
   const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
   const formattedAmount = numeral(expensesTotalAmount / 100).format('0,0[.]00 $');

   return (
      <div>
         {expensesCount !== 0 && (
            <p>
               Viewing {expensesCount} {expenseWord} totalling {formattedAmount}
            </p>
         )}
      </div>
   );
};

const mapStateToProps = (state) => {
   const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);

   return {
      expensesCount: visibleExpenses.length,
      expensesTotalAmount: getExpensesTotal(visibleExpenses),
   };
};

export default connect(mapStateToProps)(ExpensesSummary);
