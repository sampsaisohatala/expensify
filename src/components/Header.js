import { NavLink } from 'react-router-dom';

const Header = () => (
   <header>
      <h1>Expensify</h1>
      <div>
         <NavLink exact to="/" activeClassName="is-active">
            Dashboard
         </NavLink>
         <NavLink to="/create" activeClassName="is-active">
            Create Expense
         </NavLink>
      </div>
   </header>
);

export default Header;
