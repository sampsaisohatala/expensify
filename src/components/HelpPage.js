const HelpPage = () => {
   return <p>HelpPage</p>;
};

export default HelpPage;
