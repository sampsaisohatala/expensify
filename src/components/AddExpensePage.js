import React from 'react';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

const AddExpensePage = (props) => {
   return (
      <div>
         <h2>Add expense</h2>
         <ExpenseForm
            onSubmit={(expense) => {
               props.dispatch(addExpense(expense));
               props.history.push('/');
            }}
         />
      </div>
   );
};

export default connect()(AddExpensePage);
