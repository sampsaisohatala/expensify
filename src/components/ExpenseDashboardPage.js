import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
   <div>
      <h2>Dashboard</h2>
      <ExpenseListFilters />
      <ExpenseList />
   </div>
);

export default ExpenseDashboardPage;
