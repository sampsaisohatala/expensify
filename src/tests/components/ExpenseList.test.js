import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { ExpenseList } from '../../components/ExpenseList';
import expensesTestData from '../fixtures/expenses';

Enzyme.configure({
   adapter: new Adapter(),
});

test('should render ExpenseList with expenses', () => {
   const wrapper = shallow(<ExpenseList expenses={expensesTestData} />);
   expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should render ExpenseList with empty messsage', () => {
   const wrapper = shallow(<ExpenseList expenses={[]} />);
   expect(toJSON(wrapper)).toMatchSnapshot();
});
