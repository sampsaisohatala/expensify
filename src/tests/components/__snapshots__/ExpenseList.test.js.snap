// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty messsage 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Coffee"
    id="id1"
    key="id1"
    note=""
  />
  <ExpenseListItem
    amount={295000}
    createdAt={345600000}
    description="Rent"
    id="id2"
    key="id2"
    note="note"
  />
  <ExpenseListItem
    amount={6000}
    createdAt={1036800000}
    description="Car"
    id="id3"
    key="id3"
    note=""
  />
</div>
`;
